### ISSUE TITLE
State update triggered by onEndReached causes render conflict

### ISSUE THREAD

**Issue by natedevxyz** [LegendApp/legend-list#45]
When triggering a React Query fetch for the next page via the onEndReached callback in a component, I get the following error:
`ERROR  Warning: Cannot update a component (`$View`) while rendering a different component (`ForwardRef(LegendListInner2)`). To locate the bad setState() call inside `ForwardRef(LegendListInner2)`, follow the stack trace as described in https://react.dev/link/setstate-in-render`

LegendList component:
```
<LegendList
          data={notificationsData}
          renderItem={renderNotification}
          keyExtractor={(item) => item.most_recent_timestamp.toString()}
          recycleItems={false}
          initialNumContainers={8}
          showsVerticalScrollIndicator={false}
          estimatedItemSize={100}
          onEndReached={handleEndReached}
          onEndReachedThreshold={0.3}
          ListFooterComponent={renderFooter}
          refreshControl={refreshControl}
          drawDistance={480}
        />
```

handleEndReached function:
```
const handleEndReached = useCallback(() => {
    if (!isFetchingNextPage && hasNextPage) {
      fetchNextPage();
    }
  }, [isFetchingNextPage, hasNextPage, fetchNextPage]);
```

I tried delaying the fetch with a promise, requestAnimationFrame and also with a state and useEffect to wait for the render cycle but I'm still getting the error. Additionally, I get infinite fetches occurring every second after the warning is logged.

Dependencies:
"react-native": "0.76.2"
"expo": "^52.0.7"
"@tanstack/react-query": "^5.62.2"
"@legendapp/list": "^0.5.7"

**Comment by jmeistrich** [comment]
I just released version 0.5.8 which fixes some issues with onEndReached, maybe it works better now? If not, I think we may have some logic wrong. It's currently resetting onEndReached when the data changes, but that would then trigger onEndReached again... So that's probably the wrong thing to do 😅

That warning isn't really a problem so you can ignore it for now. I don't believe it actually causes any real problems, just a warning. 0.5.8 should reduce the frequency of it occurring but I'll fix it completely before a 1.0.

**Comment by ChronoByteCosmonaut** [comment]
I get the same error :  (NOBRIDGE) ERROR  Warning: Cannot update a component (`Container`) while rendering a different component (`ForwardRef(LegendListInner2)`). To locate the bad setState() call inside `ForwardRef(LegendListInner2)`, follow the stack trace as described in https://react.dev/link/setstate-in-render in LegendListInner2.

**Comment by natedevxyz** [comment]
> I just released version 0.5.8 which fixes some issues with onEndReached, maybe it works better now? If not, I think we may have some logic wrong. It's currently resetting onEndReached when the data changes, but that would then trigger onEndReached again... So that's probably the wrong thing to do 😅
> 
> That warning isn't really a problem so you can ignore it for now. I don't believe it actually causes any real problems, just a warning. 0.5.8 should reduce the frequency of it occurring but I'll fix it completely before a 1.0.

Okay, I’ll ignore the message then. What about the multiple triggers of onEndReached? Do you want me to open a new issue for that? I was able to achieve consistency by setting the onEndReachedThreshold to 0, using refs and throttling inside the handler function, and only triggering the fetch with a useEffect to match the render cycle. It feels very hacky, but it’s working.

**Comment by jmeistrich** [comment]
@natedevxyz Yes, that would be great. If it's still not working right in 0.5.8 then it's definitely an issue. I've actually not used that feature in other lists before so I'm not 100% what the correct behavior looks like... 

If you want to dig into it and PR a fix, that would be great! I'm off for the holidays and then need to catch up on stuff at work for a few days 😅

I think this line is the place to start, it's resetting whether it should call `onEndReached` again whenever data changes. I think the original reason I put it there didn't actually make sense, so removing it might just fix everything for you.

https://github.com/LegendApp/legend-list/blob/8f2447f775b5f76793eb7945fd8a0f446e7de0e0/src/LegendList.tsx#L580

**Comment by natedevxyz** [comment]
Yes, my comment was made after trying version 0.5.8.

No worries, I’ll take a look when I have time in the next couple of days. If I manage to get it working, I’ll open a PR. Thank you for pointing me in the right direction!

**Comment by thiagobrez** [comment]
I also see this warning when adding new items to the list, not related to the `onEndReached` callback.

Repro here: https://github.com/thiagobrez/legend-list-repro, in the `app/(tabs)/index.tsx` file

**Comment by jmeistrich** [comment]
Thanks @thiagobrez. It's not really a problem so you can safely ignore it, and I'll fix it soon. And I renamed this thread to be about onEndReached, and started a new one for that warning: https://github.com/LegendApp/legend-list/issues/49

**Comment by jmeistrich** [comment]
#56 was released in 0.6.2. So is this issue fixed now? :)

**Comment by natedevxyz** [comment]
I tested it locally with an infinite scrolling component. It would be nice if someone else could confirm.

**Comment by natedevxyz** [comment]
I tested it locally with an infinite scrolling component. I'll close this for now, but if someone encounters this issue again, feel free to reopen it.
